### PR TITLE
[ios] ensure ErrorRecoveryManager/Kernel can handle nil experienceId

### PR DESCRIPTION
**Pull request by esamelson**
# Why

fixes iOS side of #10598 , supersedes https://github.com/expo/expo/pull/10629

When calling reloadAsync immediately after app load, the `experienceId` is nil, but one dispatch to the main thread (for the error recovery manager) tries to access it and use it as an NSDictionary key, which causes the crash.

# How

Looking into this some more, `EXKernelAppRecord.experienceId` is explicitly marked as nullable. Almost every method in `EXErrorRecoveryManager` handles nil experienceId parameters, except for this one function. I've added an assertion that the parameter is nonnull to match what other methods in this class do, and, to be safe, I've also added null checks around all the call sites of `EXKernelAppRecord.experienceId` that did not already explicitly handle the null case.

Almost all of the places this property is accessed are for calls into the ErrorRecoveryManager, so I don't think there will be any significant issues with the property being null, as long as the methods all handle this case and don't (for example) try to use it as a dictionary key(!). I've added assertions & checks to be safe.

In other words, I don't think there are any other hidden bugs here (other than possibly ErrorRecovery missing errors with immediate reloads, which doesn't seem worth worrying about) and as long as future code that accesses `EXKernelAppRecord.experienceId` respects its `_Nullable` annotation then we should be fine to support this case.

# Test Plan

Tested an app similar to the example in #10598 that would reload immediately after loading, and set breakpoints to verify the reloading happened correctly in the following cases:

- development
- production, served from XDL
- production, published (https://expo.io/@esamelson/projects/reload-test)

### CODE REVIEW

File: ios/Exponent/Kernel/ReactAppManager/EXReactAppManager.mm [hunk 1]
```diff
@@ -674,11 +681,18 @@ - (NSDictionary * _Nullable)initialPropertiesForRootView
   NSMutableDictionary *props = [NSMutableDictionary dictionary];
   NSMutableDictionary *expProps = [NSMutableDictionary dictionary];
 
-  NSDictionary *errorRecoveryProps = [[EXKernel sharedInstance].serviceRegistry.errorRecoveryManager developerInfoForExperienceId:_appRecord.experienceId];
-  if ([[EXKernel sharedInstance].serviceRegistry.errorRecoveryManager experienceIdIsRecoveringFromError:_appRecord.experienceId]) {
-    [[EXKernel sharedInstance].serviceRegistry.errorRecoveryManager increaseAutoReloadBuffer];
-    if (errorRecoveryProps) {
-      expProps[@"errorRecovery"] = errorRecoveryProps;
+  if (_appRecord.experienceId) {
```
Comment (ide) [inline ios/Exponent/Kernel/ReactAppManager/EXReactAppManager.mm:684]: Instead of adding this guard, could we instead assert that the experienceId is nonnull? What makes me think this might be correct is that even if an app does a reload on the initial render, we already have the initial props for the root view. The one case I'm not sure about is if an app tries to reload when the update is first evaluated, before the first render:

```ts
import * as Updates from 'expo-updates';
export default Root() { ... render ... }
Updates.reloadAsync();
```

**Review by ide — APPROVED** [review]

File: ios/Exponent/Kernel/ReactAppManager/EXReactAppManager.mm (hunk 1 above)
Comment (esamelson) [reply ios/Exponent/Kernel/ReactAppManager/EXReactAppManager.mm:684]: Yep, this makes sense, and is doubly safe because these particular methods can handle a nil experienceId anyway. Just for kicks I tested this case, and did indeed get an infinite reload loop as expected (rather than a crash).
